src/metabase/search.clj

```diff
@@ -0,0 +1,118 @@
+(ns metabase.search
+  (:require [clojure.core.memoize :as memoize]
+            [clojure.string :as str]
+            [metabase.models.table :refer [Table]]
+            [schema.core :as s]))
+
+;;; Utility functions
+
+(s/defn normalize :- s/Str
+  [query :- s/Str]
+  (str/lower-case query))
+
+(s/defn tokenize :- [s/Str]
+  "Break a search `query` into its constituent tokens"
+  [query :- s/Str]
+  (filter seq
+          (str/split query #"\s+")))
+
+(def ^:private largest-common-subseq-length
+  (memoize/fifo
+   (fn
+     ([eq xs ys]
+      (largest-common-subseq-length eq xs ys 0))
+     ([eq xs ys tally]
+      (if (or (zero? (count xs))
+              (zero? (count ys)))
+        tally
+        (max
+         (if (eq (first xs)
+                 (first ys))
+           (largest-common-subseq-length eq (rest xs) (rest ys) (inc tally))
+           tally)
+         (largest-common-subseq-length eq xs (rest ys) 0)
+         (largest-common-subseq-length eq (rest xs) ys 0)))))
+   ;; Uses O(n*m) space with k < 2, so this gives us caching for at least a 22*22 search (or 50*10, etc) which sounds
+   ;; like more than enough. Memory is cheap and the items are small, so we may as well skew high
+   :fifo/threshold 1000))
+
+;;; Model setup
+
+(defn- model-name->class
+  [model-name]
+  (Class/forName (format "metabase.models.%s.%sInstance" model-name (str/capitalize model-name))))
+
+(defmulti searchable-columns-for-model
+  "The columns that will be searched for the query."
+  {:arglists '([model])}
+  class)
+
+(defmethod searchable-columns-for-model :default
+  [_]
+  [:name])
+
+(defmethod searchable-columns-for-model (class Table)
+  [_]
+  [:name
+   :display_name])
+
```

Comment:
What about `Field` as well?  I believe it has `:name` and `:display_name`?

Reply:
We don't search `Field` s at all right now. There's [an open issue](https://github.com/metabase/metabase/issues/7019), but before we tackle it we need to figure out what clicking a `Field` result should even take you to (especially if you're not an admin).

Kyle is brainstorming on this—the use cases for searching for `Field` s may well be solved by some other feature